cmd/kubeadm/app/phases/certs/renewal/expiration.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package renewal
+
+import (
+	"crypto/x509"
+	"fmt"
+	"io"
+	"math"
+	"strings"
+	"text/tabwriter"
+	"time"
+)
+
+// RenewalPeriodDays defines the leng of a period before the certificate expiration when renewal should be
+// planned and executed. This period consider that the default duration for kubeadm certificates is 365d,
+// the fact that kubeadm handles automatic certificate renewal during upgrades, and the release frequence of K8s.
+const RenewalPeriodDays = 180
```

Comment:
ideally the constant for total time (365d) and this constant should be in the same file.

but i have a bit of concern here, as we are imposing a recommendation that not all users might agree with - 180d.

i think that if the user call `check-expiration` we should always print all certs that we manage and next to them in how much time they will expire. depending on a threshold of say 1 month we can tag some of the certs as recommended for renewal.

for `upgrade plan` we can only print the certs that are are about to expire in 1 month remaining.

this would be the UX i would like to see from kubeadm if i manage my certs with it.
WDYT @fabriziopandini

Reply:
>1 Month

I have no strong opinion and the threshold is a constant and it can be changed easily

However, based on issues and slack request of help, IMO 1 month is too short, especially considering users that are going to rely on kubeadm upgrade for automatic certs renewal.

In that case, we should consider the K8s release cycle of 90d, and according to the kubeadm survey, the fact that not all the user are upgrading frequently (hence the initial proposal of 180d).

Said that I will change to whatever value there is agreement on.

> I think that if the user calls check-expiration we should always print all certs that we manage and next to them in how much time they will expire. depending on a threshold of say _X_ month we can tag some of the certs as recommended for renewal.

> for upgrade plan we can only print the certs that are are about to expire in _X_ month remaining.

This is exactly how it works; see example above from the check-expiration command; upgrade output is the same but it contains only certificates set for expiration

Reply:
> Said that I will change to whatever value there is agreement on.

ok, let's discuss tomorrow (today):

Reply:
In the meantime 

> ideally the constant for total time (365d) and this constant should be in the same file.

was fixed

Reply:
I agree that 180 days is pretty far out. 
Remember that the old rotated out cert is still valid. 
so the old cert is a liability for 180 days if we go this route. 30 or 60 days would be better.

Reply:
@mauilion this was discussed again in kubeadm office hours, and it was decided to drop the idea and always renew during upgrades